File: src/swarf/sweep.py
"""swarf sweep — move a file into .swarf/links/ and symlink it back."""

from __future__ import annotations

import shutil
from pathlib import Path

from swarf.console import error, info, warn
from swarf.exclude import add_linked_excludes
from swarf.paths import find_host_root, links_dir


def run_sweep(paths: tuple[str, ...] | list[str], host_root: Path | None = None) -> None:
    """Move files into .swarf/links/ and replace with symlinks."""
    if host_root is None:
        host_root = find_host_root()
    if host_root is None:
        error("Not inside a swarf project. Run 'swarf init' first.")
        raise SystemExit(1)

    ld = links_dir(host_root)
    if not ld.is_dir():
        error(".swarf/links/ does not exist. Run 'swarf init' first.")
        raise SystemExit(1)

    swept: list[str] = []

    for path_str in paths:
        source = Path(path_str)
        if not source.is_absolute():
            source = Path.cwd() / source

        # Check symlink before resolving
        if source.is_symlink():
            relative = source.relative_to(host_root) if source.is_relative_to(host_root) else source
            warn(f"{relative} is already a symlink, skipping.")
            continue

        # Check if inside .swarf/ BEFORE resolving (resolving follows symlinks)
        try:
            raw_relative = source.relative_to(host_root)
            if ".swarf" in raw_relative.parts:
                error(f"{path_str} is already inside .swarf/.")
                continue
        except ValueError:
            pass

        source = source.resolve()

        # Must be inside the host root
        try:
            relative = source.relative_to(host_root)
        except ValueError:
            error(f"{path_str} is not inside the project root.")
            continue

        if not source.exists():
            error(f"{path_str} does not exist.")
            continue

        # Destination in .swarf/links/
        dest = ld / relative
        dest.parent.mkdir(parents=True, exist_ok=True)

        if dest.exists():
            warn(f"{relative} already exists in .swarf/links/, skipping.")
            continue

        # Move file, create symlink
        shutil.move(str(source), str(dest))
        source.symlink_to(dest)
        swept.append(str(relative))
        info(f"  swept {relative}")

    # Update .git/info/exclude for all swept files
    if swept:
        add_linked_excludes(host_root, swept)

File: internal/sweep/sweep.go
package sweep

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/mschulkind-oss/swarf/internal/console"
	"github.com/mschulkind-oss/swarf/internal/exclude"
	"github.com/mschulkind-oss/swarf/internal/paths"
)

var (
	ErrNoProject = errors.New("not inside a swarf project — run 'swarf init' first")
	ErrNoLinks   = errors.New("swarf links/ does not exist — run 'swarf init' first")
)

func Run(filePaths []string, hostRoot string) error {
	if hostRoot == "" {
		hostRoot = paths.FindHostRoot("")
	}
	if hostRoot == "" {
		return ErrNoProject
	}
	ld := paths.LinksDir(hostRoot)
	if !paths.IsDir(ld) {
		return ErrNoLinks
	}

	var swept []string
	for _, p := range filePaths {
		if rel, ok := sweepOne(p, hostRoot, ld); ok {
			swept = append(swept, rel)
		}
	}

	if len(swept) > 0 {
		exclude.AddLinkedExcludes(hostRoot, swept)
	}
	return nil
}

func sweepOne(pathStr, hostRoot, linksDir string) (string, bool) {
	source := pathStr
	if !filepath.IsAbs(source) {
		cwd, _ := os.Getwd()
		source = filepath.Join(cwd, source)
	}

	// Check symlink before resolving
	if fi, err := os.Lstat(source); err == nil && fi.Mode()&os.ModeSymlink != 0 {
		rel, _ := filepath.Rel(hostRoot, source)
		console.Warn(fmt.Sprintf("%s is already a symlink, skipping.", rel))
		return "", false
	}

	// Check if inside swarf dir BEFORE resolving
	if rel, err := filepath.Rel(hostRoot, source); err == nil {
		for _, part := range strings.Split(rel, string(filepath.Separator)) {
			if part == paths.SwarfDirName {
				console.Error(fmt.Sprintf("%s is already inside %s/.", pathStr, paths.SwarfDirName))
				return "", false
			}
		}
	}

	resolved, err := filepath.EvalSymlinks(source)
	if err == nil {
		source = resolved
	}

	rel, err := filepath.Rel(hostRoot, source)
	if err != nil || strings.HasPrefix(rel, "..") {
		console.Error(fmt.Sprintf("%s is not inside the project root.", pathStr))
		return "", false
	}

	if _, err := os.Stat(source); os.IsNotExist(err) {
		console.Error(fmt.Sprintf("%s does not exist.", pathStr))
		return "", false
	}

	dest := filepath.Join(linksDir, rel)
	if _, err := os.Stat(dest); err == nil {
		console.Warn(fmt.Sprintf("%s already exists in %s/.links/, skipping.", rel, paths.SwarfDirName))
		return "", false
	}

	os.MkdirAll(filepath.Dir(dest), 0o755)
	if err := os.Rename(source, dest); err != nil {
		console.Error(fmt.Sprintf("Failed to move %s: %v", rel, err))
		return "", false
	}
	relTarget, err := filepath.Rel(filepath.Dir(source), dest)
	if err != nil {
		relTarget = dest // fallback to absolute
	}
	if err := os.Symlink(relTarget, source); err != nil {
		console.Error(fmt.Sprintf("Failed to create symlink for %s: %v", rel, err))
		return "", false
	}
	console.Infof("  swept %s", rel)
	return rel, true
}
